File: 100-Same-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function (p, q) {
    // Breadth First Search
    queue = [[p, q]]
    while (queue.length > 0) {
        curElem = queue.shift();

        if (curElem[0] !== null && curElem[1] !== null) {
            // Compare children value
            if (curElem[0].val !== curElem[1].val) { return false; }

            // If both tree's left or right nodes are not null,
            // add them into queue.
            if (curElem[0].left !== null && curElem[1].left !== null) {
                queue.push([curElem[0].left, curElem[1].left])
            } else if (curElem[0].left !== null || curElem[1].left !== null) {
                return false;
            }

            if (curElem[0].right !== null && curElem[1].right !== null) {
                queue.push([curElem[0].right, curElem[1].right])
            } else if (curElem[0].right !== null || curElem[1].right !== null) {
                return false;
            }
        } else {
            if (curElem[0] !== curElem[1]) { return false; }
        }
    }

    // If it couldn't find difference during traversal, two trees are same.
    return true;
};

// Sample Testcase:
//   Input:
//     [1,2,3], [1,2,3]
//   Output:
//     true

// Performance Result:
//   Coding Time: 00:30:00
//   Runtime: 84 ms, faster than 8.94% of JavaScript online submissions for Same Tree.
//   Memory Usage: 33.9 MB, less than 60.00% of JavaScript online submissions for Same Tree.

File: 100-Same-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False

        p_opened = [p]
        q_opened = [q]
        while len(p_opened) > 0:
            cur_p = p_opened.pop(0)
            cur_q = q_opened.pop(0)

            if cur_p.val != cur_q.val:
                return False

            if cur_p.left is not None and cur_q.left is not None:
                p_opened.insert(0, cur_p.left)
                q_opened.insert(0, cur_q.left)
            elif cur_p.left is None and cur_q.left is None:
                pass
            else:
                return False

            if cur_p.right is not None and cur_q.right is not None:
                p_opened.insert(0, cur_p.right)
                q_opened.insert(0, cur_q.right)
            elif cur_p.right is None and cur_q.right is None:
                pass
            else:
                return False

        return True


# Sample test case:
#   Input:
#       See the site
#   Output:
#       See the site

# Performance Result:
#   Coding Time: 00:06:41
#   Time Complexity: O(N)
#   Space Complexity: O(N)
#   Runtime: 32 ms, faster than 29.67%
#   Memory Usage: 13.8 MB, less than 5.72%
